### PR TITLE
Merge upstream shader support (r4319)

### PR DESCRIPTION
**Pull request by dreamer**
paging @krcroft @ant-222 @BPaden @James-F2

Usually, when merging upstream changes I don't do a PR/review, as there's no point. This time it's different. This PR is not for code review (unless we'll start piling fixes in here), but rather for testing and identifying regressions.

This will close issues: #100, #152 
Issues, that might've been affected and need retesting: #104 

Changes introduced in SVN r4319 were quite massive, and resulted in multiple conflicts. Git handled most of them automatically (hurray), but changes in sdlmain could not be automatically resolved -  due  to SDL2 migration. It took me some time to resolve the conflicts and results seem "ok", but we need to do thorough testing.

I left a few "TODOs" for myself during conflict resolution, where I need to spend more time, but it seems like shaders are working now. When testing (briefly) I found at least 2 regressions.

I think we should focus on functional regressions, and not on quality of shader implementation (except `glshader=sharp`, which needs to work stellar).

When testing, remember that shaders are working only with `output=opengl`; please test how various shaders interact with options: `render.aspect`, `render.scaler` (including `none`!), `sdl.fullscreen` and whatever else you'll think off.

### CODE REVIEW

**Comment by dreamer** [comment]
Ah, forgot to mention - for testing use snapshots from branch `po/test-merge-r4319` (or compile the branch yourself). :)

**Comment by dreamer** [comment]
One of the regressions I found:

## bug 1
Using:
```
[sdl]
output = opengl

[render]
aspect    = true
scaler    = none
glshader  = sharp
```
Runing PC Player Benchmark: `PCPBENCH.EXE 175`; window is stretched vertically (both in fullscreen and in windowed mode)
Screenshot from 2020-02-11 15-47-57

In other modes, window is stretched horizontally.

**Comment by dreamer** [comment]
Note for myself; relevant issues detected via PVS-Studio:
```
--- pvs-old.txt	2020-02-11 16:49:03.759801883 +0100
+++ pvs-new.txt	2020-02-11 17:00:23.905657976 +0100
@@ -4 +4 @@
-Date:	Mon Feb 10 02:55:15 2020
+Date:	Tue Feb 11 13:44:20 2020
@@ -6 +6 @@
-Project version:	e74e5240
+Project version:	e832bd97
@@ -8,3 +8,3 @@
-Command Line:	plog-converter -a GA\:1,2\;64\:1\;OP\:1,2,3\;CS\:1 -d V1042 -t fullhtml -p dosbox-staging -v e74e5240 -o pvs-report/pvs-report-2020-02-10_T0252-e74e5240 pvs-analysis.log
-Total Warnings (GA):	291
-Total Warnings (OP):	193
+Command Line:	plog-converter -a GA\:1,2\;64\:1\;OP\:1,2,3\;CS\:1 -d V1042 -t fullhtml -p dosbox-staging -v e832bd97 -o pvs-report/pvs-report-2020-02-11_T1342-e832bd97 pvs-analysis.log
+Total Warnings (GA):	293
+Total Warnings (OP):	195
@@ -238,2 +238,4 @@
-General Analysis 	sdlmain.cpp:1951 	Medium 	V779 	Unreachable code detected. It is possible that an error is present.
-General Analysis 	sdlmain.cpp:2153 	Medium 	V506 	Pointer to local variable 'myconf' is stored outside the scope of this variable. Such a pointer will become invalid.
+General Analysis 	sdlmain.cpp:2275 	Medium 	V779 	Unreachable code detected. It is possible that an error is present.
+General Analysis 	sdlmain.cpp:2477 	Medium 	V506 	Pointer to local variable 'myconf' is stored outside the scope of this variable. Such a pointer will become invalid.
+General Analysis 	sdlmain.cpp:702 	Medium 	V576 	Incorrect format. Consider checking the second actual argument of the 'GFX_ShowMsg' function. Under certain conditions the pointer can be null.
+General Analysis 	sdlmain.cpp:966 	Medium 	V576 	Incorrect format. Consider checking the second actual argument of the 'GFX_ShowMsg' function. Under certain conditions the pointer can be null.
@@ -392,0 +395 @@
+Micro-optimizations 	render.cpp:612 	Low 	V807 	Decreased performance. Consider creating a reference to avoid using the 'buf.str()' expression repeatedly.
@@ -416,0 +420 @@
+Micro-optimizations 	sdlmain.cpp:222 	Low 	V802 	On 64-bit platform, structure size can be reduced from 112 to 104 bytes by rearranging the fields according to their sizes in decreasing order.
@@ -568 +572 @@
-Customers Specific 	render.cpp:259 	High 	V2008 	Cyclomatic complexity: 66. Consider refactoring the 'RENDER_Reset' function.
+Customers Specific 	render.cpp:264 	High 	V2008 	Cyclomatic complexity: 66. Consider refactoring the 'RENDER_Reset' function.
@@ -595,2 +599,2 @@
-Customers Specific 	sdlmain.cpp:1219 	High 	V2008 	Cyclomatic complexity: 92. Consider refactoring the 'GUI_StartUp' function.
-Customers Specific 	sdlmain.cpp:580 	High 	V2008 	Cyclomatic complexity: 67. Consider refactoring the 'GFX_SetSize' function.
+Customers Specific 	sdlmain.cpp:1518 	High 	V2008 	Cyclomatic complexity: 109. Consider refactoring the 'GUI_StartUp' function.
+Customers Specific 	sdlmain.cpp:727 	High 	V2008 	Cyclomatic complexity: 83. Consider refactoring the 'GFX_SetSize' function.
```

File: src/gui/render.cpp [hunk 1]
```diff
@@ -609,8 +609,8 @@ static bool RENDER_GetShader(std::string& shader_path) {
 	else if (shader_path == "tv3x")        buf << tv3x_glsl;
 	else if (shader_path == "sharp")       buf << sharp_glsl;
 
-	if (!buf.str().empty()) {
-		std::string s = buf.str();
+	std::string s = buf.str();
```
Comment (krcroft) [inline src/gui/render.cpp]: Is `s` a reference or copy?  (Just noticed that further down we update `s`, so probably need to be reference).

**Comment by BPaden** [comment]
This doesn't go quite so far as implementing #100, which includes compatibility with the numerous retroarch shaders and shader-preset files. But for the purposes of dosbox, I think this is probably enough for most people if we can get crt-lottes ported over.

Anyway, I tested with Doom, Quake, Descent, Keen 4 and magiduck with no issues noticed. This is all with scaler=none, glshader=sharp and aspect=true. Does it even make sense to use scaler and glshader at the same time?

**Comment by dreamer** [comment]
> This doesn't go quite so far as implementing #100, which includes compatibility with the numerous retroarch shaders and shader-preset files. But for the purposes of dosbox, I think this is probably enough for most people if we can get crt-lottes ported over.
> 
> Anyway, I tested with Doom, Quake, Descent, Keen 4 and magiduck with no issues noticed. This is all with scaler=none, glshader=sharp and aspect=true. Does it even make sense to use scaler and glshader at the same time?

@BPaden Understood, thank you for testing :)

When merging this in, I will keep #100 open for the purpose of supporting shaders in retroarch format and presets (GLSL only). I don't know Slang shader language, so I can't comment about that one.

**Comment by dreamer** [comment]
@BPaden 
> This is all with scaler=none, glshader=sharp and aspect=true. Does it even make sense to use scaler and glshader at the same time?

I don't think so. However when a different scaler is selected, initial window size changes - so I'm not sure if author of shader code upscales raw input, or input already upscaled once via software scalers.

Unfortunately, there's no documentation for this in SVN and commit message is extremely brief for such a large change, I need to finish reading the code.

**Comment by dreamer** [comment]
Turns out **bug 1** I mentioned earlier is not a regression - the bug is reproducible in DOSBox 0.74-3.

**Comment by dreamer** [comment]
Merged r4319 and r4320 to master branch; we'll deal with regressions as new issues.

**Comment by krcroft** [comment]
Thanks for making that call @dreamer.

**Pull request closed without merging**